lib/stages/visSharedMemWriter.cpp

```diff
@@ -0,0 +1,393 @@
+        FATAL_ERROR("Failed to release semaphore {}", sem_name);
+        return;
+    }
+
+    // first write the metadata, then the data, then the valid byte
+    // add valid_size amount of padding
+    memcpy(buf_write_pos + valid_size, frame.metadata(), metadata_size);
+    memcpy(buf_write_pos + metadata_size + valid_size, frame.data(), data_size);
+    memcpy(buf_write_pos, &valid, sizeof(valid));
+
+    // Document the fpga sequence counter for that frame in the access record
+    uint64_t fpga_seq = frame.metadata()->fpga_seq_start;
+
+
+    if (sem_wait(sem) == -1) {
+        FATAL_ERROR("Failed to acquire semaphore {}", sem_name);
+        return;
+    }
+
+    INFO("Writing fpga_seq {} to index {}", fpga_seq, time_ind);
+    memcpy(access_record_write_pos, &fpga_seq, sizeof(fpga_seq));
+
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", sem_name);
+        return;
+    }
+
+    // update num_writes
+    num_writes++;
+    memcpy(structured_data_addr, &num_writes, sizeof(num_writes));
+}
+
+uint64_t visSharedMemWriter::get_data_size(const visFrameView& frame) {
+
+    auto& dm = datasetManager::instance();
+
+    // Get properties of stream from first frame and datasetManager
+    // If we can get the data_size in other ways, we will not need
+    // ninput, nvis, or num_ev anymore
+    auto sstate_fut = std::async(&datasetManager::dataset_state<stackState>, &dm, frame.dataset_id);
+    auto istate_fut = std::async(&datasetManager::dataset_state<inputState>, &dm, frame.dataset_id);
+    auto pstate_fut = std::async(&datasetManager::dataset_state<prodState>, &dm, frame.dataset_id);
+
+    auto evstate_fut = std::async(&datasetManager::dataset_state<eigenvalueState>, &dm, frame.dataset_id);
+
+    const inputState* istate = istate_fut.get();
+    const prodState* pstate = pstate_fut.get();
+    const stackState* sstate = sstate_fut.get();
+    const eigenvalueState* evstate = evstate_fut.get();
+
+
+    if (!istate || !pstate) {
+        ERROR("Required datasetState not found for dataset ID {}\nThe following required states "
+                "were found:\ninputState - {:p}\nprodState - {:p}\n",
+                frame.dataset_id, (void*)istate, (void*)pstate);
+        throw std::runtime_error("Could not write to shared memory.");
+    }
+
+    // Count the eigenvalue index
+    size_t num_ev;
+    if (evstate) {
+        num_ev = evstate->get_num_ev();
+    } else {
+        num_ev = 0;
+    }
+
+    size_t ninput = istate->get_inputs().size();
+    size_t nvis = sstate ? sstate->get_num_stack() : pstate->get_prods().size();
+
+    auto layout = visFrameView::calculate_buffer_layout(ninput, nvis, num_ev);
+
+    return layout.first;
+}
+
+void visSharedMemWriter::main_thread() {
+    INFO("Reached main thread");
+
+    frameID frame_id(in_buf);
+
+    // The current position in the ring buffer of the most recent time sample
+    // from 0 -> ntime
+    cur_pos = modulo<int>(ntime);
```

Comment:
I don't think this is equivalent to `cur_pos = 0`. Instead make `cur_pos` be of type `modulo<int>` (or `modulo<size_t>`?) in `visSharedMemWriter.hpp`

Reply:
note that you have to cast it back to int in line 111